pkg/validators/quota.go

```diff
@@ -216,3 +218,56 @@ func serviceLimits(consumer string, service string) ([]*sub.ConsumerQuotaMetric,
 		})
 	return res, err
 }
+
+type usageKey struct {
+	Metric   string
+	Location string // either "global", region, or zone
+}
+
+type usageProvider struct {
+	u map[usageKey]int64
+}
+
+func (up *usageProvider) Usage(metric string, region string, zone string) int64 {
+	if up.u == nil {
+		return 0
+	}
+	k := usageKey{metric, "global"}
+	if region != "" {
+		k.Location = region
+	}
+	if zone != "" {
+		k.Location = zone
+	}
+	return up.u[k] // 0 if not found
+}
+
+func newUsageProvider(project_id string) (usageProvider, error) {
+	s, err := cm.NewService(context.Background())
+	if err != nil {
+		return usageProvider{}, err
+	}
+
+	u := map[usageKey]int64{}
+	err = s.Projects.TimeSeries.List("projects/"+project_id).
+		Filter(`metric.type="serviceruntime.googleapis.com/quota/allocation/usage" resource.type="consumer_quota"`).
+		IntervalEndTime(time.Now().Format(time.RFC3339)).
+		// Quota usage metrics get duplicated once a day
```

Comment:
Is this only true for metrics that haven't changed in 24 hours?  Or will we only be able to retrieve data that's up to 24 hours old?

Reply:
If metric wasn't updated with 24 hours, it will be duplicated.
It is to prevent metrics that aren't update frequently to fall of the horizon.